contracts/TaskBook.sol

```diff
@@ -5,83 +5,158 @@ import './AccountManager.sol';
 contract TaskBook is AccountManager {
 
 	uint private numTasks = 0;
-	uint constant maxSolvers = 10;
-	uint constant maxChallengers = 10;
+	uint private forcedErrorThreshold = 42;
 
 	event TaskCreated(uint taskID, uint minDeposit, uint blockNumber);
 	event SolverSelected(uint indexed taskID, address solver, bytes32 taskData, uint minDeposit);
-	event SolutionCommitted(uint taskID, uint minDeposit, bytes32 taskData, address solver);
+	event SolutionsCommitted(uint taskID, uint minDeposit, bytes32 taskData, address solver);
+	event SolutionRevealed(uint taskID, uint randomBits);
+	event TaskStateChange(uint taskID, uint state);
 
-	mapping(uint => Task) private tasks;
-	mapping(address => mapping(uint => bytes32)) private solverRandomBitsHash;
-	mapping(address => mapping(uint => bytes32)) private solutions;
-	mapping(address => mapping(uint => bytes32)) private verifierIntent;
+	struct Challenge {
+		address challenger;
+		bytes32 intentHash;
+	}
 
 	struct Task {
 		address owner;
-		address[] solvers;
 		address selectedSolver;
 		uint minDeposit;
 		bytes32 taskData;
-		uint numSolvers;
-		address[] challengers;
-		uint numChallengers;
+		mapping(address => bytes32) challenges;
+		uint state;
 	}
 
+	struct Solution {
+		bytes32 solutionHash0;
+		bytes32 solutionHash1;
+		bool solution0Correct;
+		address[] solution0Challengers;
+		address[] solution1Challengers;
+	}
+
+	mapping(uint => Task) private tasks;
+	mapping(address => mapping(uint => bytes32)) private solverRandomBitsHash;
+	mapping(uint => Solution) private solutions;
+
 	//Task Issuers create tasks to be solved
 	function createTask(uint minDeposit, bytes32 taskData, uint numBlocks) returns (bool) {
 		require(balances[msg.sender] >= minDeposit);
-		tasks[numTasks] = Task(msg.sender, new address[](maxSolvers), 0x0, minDeposit, taskData, 0, new address[](maxChallengers), 0);
+		Task storage t;
+		t.owner = msg.sender;
+		t.minDeposit = minDeposit;
+		t.taskData = taskData;
+		tasks[numTasks] = t;
 		log0(sha3(msg.sender));//possible bug if log is after event
 		TaskCreated(numTasks, minDeposit, block.number+numBlocks);
 		numTasks++;
 		return true;
 	}
 
-	//Solver registers for tasks
+	function changeTaskState(uint taskID, uint newState) returns (bool) {
+		Task storage t = tasks[taskID];
+		require(t.owner == msg.sender);
+		t.state = newState;
+		TaskStateChange(taskID, newState);
+		return true;
+	}
+
+	//Solver registers for tasks, if first to register than automatically selected solver
+	//0->1
 	function registerForTask(uint taskID, uint minDeposit, bytes32 randomBitsHash) returns(bool) {
 		require(balances[msg.sender] >= minDeposit);
-		Task t = tasks[taskID];
+		Task storage t = tasks[taskID];
 		require(!(t.owner == 0x0));
-		require(t.numSolvers < maxSolvers);
+		require(t.state == 0);
+		require(t.selectedSolver == 0x0);
+		t.selectedSolver = msg.sender;
```

Comment:
I like this. Whichever Solvers bid gets in first is automatically selected as Solver. :)